### ISSUE TITLE
Preserve WillPopScope as an alternate for PopScope

### ISSUE THREAD

**Issue by AndreiMisiukevich** [flutter/flutter#138614]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Is there any chance the decision about deprecating WillPopScope will be revisited?

My team finds PopScope inconvenient in comparison to WillPopScope.

While WillPopScope allows to intercept the pop-action intent, the PopScope doesn't. Instead, PopScope proposes to handle the result of this action.

### Proposal

In our case, we want to present confirmation dialogs when the user wants to close the page and WillPopScope works like a charm in this scenario.

Yes, we can set canPop: false and write dialogue prompting logic inside onPopInvoked, but it doesn't solve the problem because the result of NavigatorState.maybePop will be returned before the user answers to dialog.

Also, popDisposition is sync, while willPop is async. This makes prompting a dialogue impossible.

**Comment by jellynoone** [comment]
I second this proposal, for the following reason.

Sometimes, a page might want to prevent pops from multiple places. With `WillPopScope` this was trivial since if any of the `onWillPop` callback returned false, the pop would be prevented. This was especially useful for, for instance, displaying a loading overlay and have it register pop prevention callbacks.

With `PopScope` and `PopEntry` the same behaviour is quite more difficult to achieve and will no longer work out of the box, since now **ALL** the `PopScope`s and their callbacks will need to be synchronised in order to ensure there aren't duplicate pops going on by the callbacks.

In case, a user's route only contains one `PopScope` using the solution described in https://github.com/flutter/website/pull/9244 works. In case, the user wants to use multiple scopes or PopEntries, they will more or less have to reimplement the WillPop callbacks mechanisms themselves in probably a much more error prone way, since now this functionality would have to be provided by some additional widget they'd have to manage on their own.

While I don't fully understand the details of why these changes took place, the solution recommended in https://github.com/flutter/website/pull/9244 is to apply `false` to `PopScope.canPop` and then provide a handler that asks for confirmation in `onPopInvoked`.
Would it be possible to keep the `WillPopScope` and simply have a `Route` with willPopScopeCallbacks be treated as having a `PopEntry` with `canPop` set to false?

Cc: @justinmc as the author of: #132249, and @chunhtai

**Comment by chunhtai** [comment]
The reason for this change is for the new Android predictive back. Flutter app need to preemptively tell the android activity if it can pop before the pop is actually called. WillpopScope does not work in this case, unless we assume it will always return false.

**Comment by sunjiaming** [comment]
Follow up

**Comment by Kaizodo** [comment]
I got same issue , i want to ask user before closing,  with PopScope i have to render entire UI again just to update a boolean value of canPop . this is also casuing issues with reactive frameworks like GetX  . 

with WillPopScope a callback function was checking for the updated value and simply preventing the pop of activity.

**Comment by TatsuUkraine** [comment]
> I got same issue , i want to ask user before closing, with PopScope i have to render entire UI again just to update a boolean value of canPop . this is also casuing issues with reactive frameworks like GetX .
> 
> with WillPopScope a callback function was checking for the updated value and simply preventing the pop of activity.

if you want to keep PopScope working in the same way as WillPopScope works, just keep `canPop: false` always and manually call `Navigator.of(context).pop()` in `onPopInvoked` method when you need it. The only difference will be that `onPopInvoked` doesn't expect you to return bool value at the end, so it's fair to say that the following code samples work in the same way:
```dart
PopScope(
  canPop: false,
  onPopInvoked: (didPop) {
     if (someConditionToAllowBackNavigation) {
       Navigator.of(context).pop();
       return;
     }

     /// someother logic
  },
  child: child,
)
```

```dart
WillPopScope(
  onWillPop: () async {
     if (someConditionToAllowBackNavigation) {
       Navigator.of(context).pop();
     } else {
       /// someother logic
     }

     return false;
  },
  child: child,
)
```

**Note**, WillPopScope example is not the best one) I wrote it in this way to show the similarity with PopScope by making `onWillPop` always return false

**Comment by fernandooliiveira** [comment]
> > I got same issue , i want to ask user before closing, with PopScope i have to render entire UI again just to update a boolean value of canPop . this is also casuing issues with reactive frameworks like GetX .
> > with WillPopScope a callback function was checking for the updated value and simply preventing the pop of activity.
> 
> if you want to keep PopScope working in the same way as WillPopScope works, just keep `canPop: false` always and manually call `Navigator.of(context).pop()` in `onPopInvoked` method when you need it. The only difference will be that `onPopInvoked` doesn't expect you to return bool value at the end, so it's fair to say that the following code samples work in the same way:
> 
> ```dart
> PopScope(
>   canPop: false,
>   onPopInvoked: (didPop) {
>      if (someConditionToAllowBackNavigation) {
>        Navigator.of(context).pop();
>        return;
>      }
> 
>      /// someother logic
>   },
>   child: child,
> )
> ```
> 
> ```dart
> WillPopScope(
>   onWillPop: () async {
>      if (someConditionToAllowBackNavigation) {
>        Navigator.of(context).pop();
>      } else {
>        /// someother logic
>      }
> 
>      return false;
>   },
>   child: child,
> )
> ```
> 
> **Note**, WillPopScope example is not the best one) I wrote it in this way to show the similarity with PopScope by making `onWillPop` always return false

The PopScope doesn't work the same way as WillPopScope, see this example:

```
WillPopScope(
  onWillPop: () async {
    if (!canPop) {
      return false;
    }
    Navigator.of(context).pop();
    return true;
  },
);
```

While using WillPopScope my condition works perfectly, but when trying to use PopScope:

```
PopScope(
  canPop: false,
  onPopInvoked: (_) {
    if (!canPop) {
      return;
    }
    Navigator.of(context).pop();
  },
);
```

It throws this exception:

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: 'package:flutter/src/widgets/navigator.dart': Failed assertion: line 5238 pos 12: '!_debugLocked': is not true.
#0      _AssertionError._doThrowNew (dart:core-patch/errors_patch.dart:51:61)
#1      _AssertionError._throwNew (dart:core-patch/errors_patch.dart:40:5)
#2      NavigatorState.pop (package:flutter/src/widgets/navigator.dart:5238:12)
#3      _AdmFunctionsContentState.build.<anonymous closure>.<anonymous closure> (package:app/pages/home/widgets/adm_functions/adm_functions_content.dart:98:33)
#4      ModalRoute.onPopInvoked (package:flutter/src/widgets/routes.dart:1574:30)
#5      _RouteEntry.pop (package:flutter/src/widgets/navigator.dart:3083:11)
#6      NavigatorState.pop (package:flutter/src/widgets/navigator.dart:5251:13)
#7      _AdmFunctionsContentState.build.<anonymous closure>.<anonymous closure> (package:app/pages/home/widgets/adm_functions/adm_functions_content.dart:98:33)
#8      ModalRoute.onPopInvoked (package:flutter/src/widgets/routes.dart:1574:30)
#9      NavigatorState.maybePop (package:flutter/src/widgets/navigator.dart:5207:25)
<asynchronous suspension>
```

And, as already mentioned, it is necessary to reload the entire screen in case a boolean is used in the canPop parameter of PopScope.
The PopScope does not work reactively.

Note: I am creating a Desktop application.

**Comment by luis901101** [comment]
@fernandooliiveira yep, it does not work the same way, I'm also struggling somehow to getting use to the new way PopScope works.
Anyway just in case, the **Exception** you are having is because after you call `Navigator.of(context).pop();` the `onPopInvoked` function is called again and thus the `Navigator.of(context).pop();` is also called again using a context already disposed, that’s why the Exception is thrown, to avoid this you can use a simple bool flag there setting it to true before calling pop… something like: 

```dart
bool alreadyPopped= false;

void onPopInvoked: (_) {
    if(alreadyPopped) return;
    if (!canPop) {
      return;
    }
    alreadyPopped = true;
    Navigator.of(context).pop();
  };
```
Note you could use the same bool flag as the `canPop` in `PopScope`, but you don't have to `setState((){})` if you already call `Navigator.of(context).pop();`

**Comment by TatsuUkraine** [comment]
> While using WillPopScope my condition works perfectly, but when trying to use PopScope

I feel like with this variant your code will be doing 2 pop calls 🤔 since you making explicit pop call and then system will also pass forward original pop attempt since you returning `true`

**Comment by hwkim1127** [comment]
with popScope: false, when I call anything with `context` it throws the error. And I am not using third party package for navigation. It even throws error when I use `Navigator.of(context).push` to show some full page `Scaffold Widget` even without any third party libraries.

**Comment by KuroodoD** [comment]
I have a use case where I am showing a dialog, and if the user tries to exit this dialog there is a dialog to confirm whether or not the user wishes to exit. PopScope has introduced a lot of issues, and is less elegant than using a simple WillPopScope. I understand that the justification for the change is due to changes in the Android API. But Flutter is also used for other platforms not just Android.

Here is an example of what that looks like with WillPopScope:

```
return WillPopScope(
      onWillPop: () async {
        return await showDialog<bool>(
              ... 
         ) ?? false;
      },
      child: ...
```


Here it is with PopScope:

```
bool _canPop = false
...
return PopScope(
      canPop: _canPop,
      onPopInvoked: (didPop) {
        if (_canPop) return;
        WidgetsBinding.instance.addPostFrameCallback(
          (_) async {
            _canPop = await showDialog<bool>(
              ... 
           ) ?? false;

            if (_canPop) {
              if (!mounted) return;
              Navigator.of(context).pop();
            }
          },
        );
      },
      child: ...
```

PopScope is so involved compared. There's no reason for so much complexity for something so simple.

PopScope also introduces some new issues. If you click away from a dialogue, or in other words the barrier dismissible feature, popping does not get invoked. Nothing happens. But the moment you hover over a text field or a button, the pop finally gets invoked. If you click multiple times, you will end up having onPopInvoked called multiple times. In the below video, you can see that onPopInvoked is never called until I hover (not press) over the button. This issue is not present when using WillPopScope

https://github.com/flutter/flutter/assets/148470439/3f988f62-7b10-4224-a357-677a90708ebb

Here is the dart code you can use to see the problems with PopScope and dialogs:
[main.txt](https://github.com/flutter/flutter/files/13621724/main.txt) and [main_willPopScope.txt](https://github.com/flutter/flutter/files/13621725/main_willPopScope.txt)

**Comment by TatsuUkraine** [comment]
Yeah, as I said before PopScope has issues) the point I was trying to make is that maybe problem not with removing WillPopScope (or keeping it) but simply because PopScope has so many probles, that it doesn't behave as it supposed to and not all usecase were taken into account

P.s I'm not saying that it accurate assumption, just saying that maybe it just a bigger issue with new widget that should be fixed, rather than keeping both. And I personally don't have strong preference one over another

**Comment by amake** [comment]
My understanding is that using `PopScope` with `canPop: true` is nice in that it facilitates Android's predictive back gesture.

But I don't understand how I'm supposed to use it with `canPop: true` when I need to return something from my route even in the case of a back gesture. The way I know to do this is to call `Navigator.of(context).pop(returnValue)`, but the only way I see to do that universally is to have `canPop: false` and do it in the `onPopInvoked` callback.

How is this case intended to work?

**Comment by ekuleshov** [comment]
> Here it is with PopScope:
> 
> ```
> bool _canPop = false
> ...
> return PopScope(
>       canPop: _canPop,
>       onPopInvoked: (didPop) {
>         if (_canPop) return;
>         WidgetsBinding.instance.addPostFrameCallback(
>           (_) async {
>             _canPop = await showDialog<bool>(
>               ... 
>            ) ?? false;
> 
>             if (_canPop) {
>               if (!mounted) return;
>               Navigator.of(context).pop();
>             }
>           },
>         );
>       },
>       child: ...
> ```
> 
> PopScope is so involved compared. There's no reason for so much complexity for something so simple.

@KuroodoD @chunhtai and all... Is this code our current best option to ask confirmation *before* leaving the `PopScope` and then stay on the activity/screen when user dismisses it?

The `WidgetsBinding.instance.addPostFrameCallback` looks like a hack and also requires a stateful widget.

**Comment by TatsuUkraine** [comment]
@ekuleshov here is a migration guide with all scenarios 
https://docs.flutter.dev/release/breaking-changes/android-predictive-back#migrating-a-back-confirmation-dialog
This particular example shows most common usage of WillPopScope. Just keep on mind that there are still might be some bugs that aren't resolved yet with new back navigation handling when it comes to handling native back navigation on root screens

**Comment by ekuleshov** [comment]
> @ekuleshov here is a migration guide with all scenarios https://docs.flutter.dev/release/breaking-changes/android-predictive-back#migrating-a-back-confirmation-dialog This particular example shows most common usage of WillPopScope. Just keep on mind that there are still might be some bugs that aren't resolved yet with new back navigation handling when it comes to handling native back navigation on root screens

That kind of works. However it looks like you also have to duplicate the same confirmation in the `AppBar.leading` widget because the `PopScope.onPopInvoked` gets `didPop == true` after the `AppBar.leading` is tapped.

**Comment by Zeinab-Kouhkan** [comment]
I am using this  code :
`return  PopScope(
        canPop: false,
        onPopInvoked: (bool didPop) {
          debugPrint("---------------$didPop");
          if (didPop) {
            return;
          }

          final NavigatorState navigator = Navigator.of(context);
          final bool shouldPop = bloc.onWillPop();

          if (shouldPop ?? false) {
            navigator.pop();
          }
        }, child: ...,
);`


didPop always is true

**Comment by kane-knowby** [comment]
> I am also using some confirmation dialogs in my app and so far I had not many issues with PopScope...
> 
> I implemented my confirmation dialogs as follows:
> 
> ```dart
> ...
> 
> return PopScope(
>   canPop: false,
>   onPopInvoked: (didPop) async {
>     if (didPop) return;
> 
>     final close = await handleClose();
>     if (close && mounted) {
>       Navigator.of(context).pop();
>     }
>   },
>   child: ...,
> );
> ```
> 
> where `handleClose()` is just another async function that shows the confirmation dialog. Granted this is still using the `mounted` attr. of the stateful widget, but I think it should work in a stateless as well... Also I had no issues running async code in the `onPopInvoked` method. This code currently works on Flutter 3.16.5.
> 
> I'd be interested in finding out if / how my approach is wrong 😅

This was super helpful, I implemented my own reusable widget based on this answer to go with auto_route

``` dart
typedef CanGoBackCallback = Future<bool?> Function();

class BackGuard extends StatelessWidget {
  const BackGuard({
    required this.child,
    required this.canGoBack,
    super.key,
  });

  final Widget child;
  final CanGoBackCallback canGoBack;

  @override
  Widget build(BuildContext context) {
    return PopScope(
      canPop: false,
      onPopInvoked: (popped) {
        if (popped) return;
        canGoBack().then((pop) => (pop ?? false) ? context.router.back() : null);
      },
      child: child,
    );
  }
}
```

Working perfectly now thank you! @florian-obernberger

**Comment by justinmc** [comment]
Let me try to wrap up the problems in this issue and try to get people pointed in the right direction. If you're having a problem that's not covered here, please open a new issue (and feel free to tag me).

### Known bugs

 * [PopScope doesn't work with Navigator.pages](https://github.com/flutter/flutter/issues/141189)
   - Fixed in https://github.com/flutter/flutter/pull/141221.
   - Another somewhat related fix: https://github.com/flutter/flutter/pull/142190
 * It's hard to use PopScope with a confirmation dialog
   - Mentioned in this issue in the comments above.
   - Solution offered by @florian-obernberger :https://github.com/flutter/flutter/issues/138614#issuecomment-1883705722
   - I've opened a PR (https://github.com/flutter/flutter/pull/142163) to update the examples to make them more clear about this.
     - Example with Form: https://github.com/flutter/flutter/blob/master/examples/api/lib/widgets/form/form.1.dart
     - Example with PopScope: https://github.com/flutter/flutter/blob/master/examples/api/lib/widgets/pop_scope/pop_scope.0.dart
 * Problems with go_router
   - https://github.com/flutter/flutter/issues/140869
   - https://github.com/flutter/flutter/issues/138737, 
   - Waiting on PR https://github.com/flutter/flutter/pull/137792. I'm unsure if an additional change to go_router is needed.
 * Problems with PopScope and iOS
    - https://github.com/flutter/flutter/issues/141268
    - https://github.com/flutter/flutter/issues/138624
      - Also https://github.com/flutter/flutter/issues/137458
      - Fixed with PR https://github.com/flutter/flutter/pull/137792. Or WAI?
 * Can't get the result of a route in onPopInvoked
    - To be reopened and fixed in PR https://github.com/flutter/flutter/pull/139164.

### Can we bring back WillPopScope?
WillPopScope appears to be fundamentally incompatible with Android's new back architecture ([explained in the design doc](https://docs.google.com/document/d/1BGCWy1_LRrXEB6qeqTAKlk-U2CZlKJ5xI97g45U7azk/edit?tab=t.0#heading=h.xf8t3pyo7yqj)). Meanwhile, the problems with PopScope, outlined in this comment, appear to be either misunderstandings or fixable bugs. I think we should spend our effort getting everyone successfully migrated to PopScope rather than maintaining two similar APIs, one of which doesn't fully work on Android.

**Comment by georrge1994** [comment]
Instead the simple solution we got a hack. Who did decide to deprecate it? Which of clear mind?

**Comment by vijayvaghela72** [comment]
not valid solution not working proper as required

**Comment by Abhijit-Revamp** [comment]
@justinmc 
Can you please include WEB to this issue (in addition to Android and iOS). 
It's broken when using browser's BACK button (Chrome and other supported browsers).

**Comment by justinmc** [comment]
@Abhijit-Revamp Can you open a new issue for that?

**Comment by ClebsonDantasUchoa** [comment]
I created a custom PopScope to work around the problem:

`import 'package:flutter/material.dart';


class CustomPopScope extends StatefulWidget {

  final Future<bool> Function() canPop;
  final Widget child;

  const CustomPopScope({required this.canPop, required this.child, super.key});

  @override
  State<CustomPopScope> createState() => _CustomPopScopeState();
}

class _CustomPopScopeState extends State<CustomPopScope> {

  bool canClosePage = false;

  @override
  Widget build(BuildContext context) {
    if (canClosePage) {
      return widget.child;
    }
    return PopScope(
      canPop: false,
      onPopInvoked: (_) {
        widget.canPop().then((canPop) {
          if (canPop) {
            setState(() => canClosePage = true);
            Navigator.pop(context);
          }
        });
      },
      child: widget.child,
    );
  }

}`

**Comment by ClebsonDantasUchoa** [comment]
I created a custom PopScope to work around the problem.

```
import 'package:flutter/material.dart';


class CustomPopScope extends StatefulWidget {

  final Future<bool> Function() onWillPop;
  final Widget child;

  const CustomPopScope({required this.onWillPop, required this.child, super.key});

  @override
  State<CustomPopScope> createState() => _CustomPopScopeState();
}

class _CustomPopScopeState extends State<CustomPopScope> {

  bool canClosePage = false;

  @override
  Widget build(BuildContext context) {
    if (canClosePage) {
      return widget.child;
    }
    
    return PopScope(
      canPop: false,
      onPopInvoked: (_) {
        widget.onWillPop().then((canPop) {
          if (canPop) {
            setState(() => canClosePage = true);
            Navigator.pop(context);
          }
        });
      },
      child: widget.child,
    );
  }  

}
```

**Comment by ClebsonDantasUchoa** [comment]
I created a custom PopScope to work around the problem.

```
import 'package:flutter/material.dart';


class CustomPopScope extends StatefulWidget {

  final Future<bool> Function() onWillPop;
  final Widget child;

  const CustomPopScope({required this.onWillPop, required this.child, super.key});

  @override
  State<CustomPopScope> createState() => _CustomPopScopeState();
}

class _CustomPopScopeState extends State<CustomPopScope> {

  bool _usePopScope = true;

  @override
  Widget build(BuildContext context) {
    if (_usePopScope = false) {
      return widget.child;
    }

    return PopScope(
      canPop: false,
      onPopInvoked: (_) {
        widget.onWillPop().then((canPop) {
          if (canPop && _usePopScope == false) {
            setState(() => _usePopScope = false);
            Navigator.pop(context);
          }
        });
      },
      child: widget.child,
    );
  }  

}
```

**Comment by randsoy** [comment]
> I created a custom PopScope to work around the problem.
> 
> ```
> import 'package:flutter/material.dart';
> 
> 
> class CustomPopScope extends StatefulWidget {
> 
>   final Future<bool> Function() onWillPop;
>   final Widget child;
> 
>   const CustomPopScope({required this.onWillPop, required this.child, super.key});
> 
>   @override
>   State<CustomPopScope> createState() => _CustomPopScopeState();
> }
> 
> class _CustomPopScopeState extends State<CustomPopScope> {
> 
>   bool _usePopScope = true;
> 
>   @override
>   Widget build(BuildContext context) {
>     if (_usePopScope = false) {
>       return widget.child;
>     }
> 
>     return PopScope(
>       canPop: false,
>       onPopInvoked: (_) {
>         widget.onWillPop().then((canPop) {
>           if (canPop && _usePopScope == false) {
>             setState(() => _usePopScope = false);
>             Navigator.pop(context);
>           }
>         });
>       },
>       child: widget.child,
>     );
>   }  
> 
> }
> ```

import 'package:flutter/material.dart';

class MuPopScope extends StatefulWidget {
  final Future<bool> Function() onWillPop;
  final Widget child;

  const MuPopScope({required this.onWillPop, required this.child, super.key});

  @override
  State<MuPopScope> createState() => _MuPopScopeState();
}

class _MuPopScopeState extends State<MuPopScope> {
  bool _usePopScope = true;

  @override
  Widget build(BuildContext context) {
    if (_usePopScope == false) {
      return widget.child;
    }

    return PopScope(
      canPop: false,
      onPopInvoked: (_) {
        widget.onWillPop().then((canPop) {
          if (canPop && _usePopScope) {
            setState(() => _usePopScope = false);
            Navigator.pop(context);
          }
        });
      },
      child: widget.child,
    );
  }
}

**Comment by randsoy** [comment]
@ClebsonDantasUchoa 
```
import 'package:flutter/material.dart';

class MuPopScope extends StatefulWidget {
  final Future<bool> Function() onWillPop;
  final Widget child;

  const MuPopScope({required this.onWillPop, required this.child, super.key});

  @override
  State<MuPopScope> createState() => _MuPopScopeState();
}

class _MuPopScopeState extends State<MuPopScope> {
  bool _usePopScope = true;

  @override
  Widget build(BuildContext context) {
    if (_usePopScope == false) {
      return widget.child;
    }

    return PopScope(
      canPop: false,
      onPopInvoked: (_) {
        widget.onWillPop().then((canPop) {
          if (canPop && _usePopScope) {
            setState(() => _usePopScope = false);
            Navigator.pop(context);
          }
        });
      },
      child: widget.child,
    );
  }
}

```

**Comment by kamaravichow** [comment]
I guess you can just use PopScope instead, here is the guide https://www.aravi.me/blog/willpopscope-is-now-popscope-in-flutter/

**Comment by kamaravichow** [comment]
Github issues became new stack overflow, I just linked the article for beginners who need more explaination. And yeah I wanted people to find what I write.

**Comment by mkucharski17** [comment]
I have a problem with depreciation of `WillPopScope`. I am using it to prevent app closing on native back. Workaround which is here requires `WillPopScope`:  https://github.com/flutter/flutter/issues/66349#issuecomment-696788657.

**Comment by justinmc** [comment]
Not sure about your specific case, but in general you can prevent a system back from closing the app with `PopScope(canPop: false, ...)`. You should open an issue if you've got a bigger problem or you suspect a bug in PopScope.

**Comment by topex-psy** [comment]
I've subscribed this issue for a long time, still hoping for the light to come.

Can anyone or Flutter team itself explain how to migrate from `WillPopScope` to `PopScope` in a video of anything which is easy to understand?

Or is it really needed to deprecate `WillPopScope`? Are they cannot both co-exist? Why?

I used many `WillPopScope` in my single published project. I've tried several times to try to migrate, but I'm afraid I'm not doing it right & break my already published working app on Play Store.

One of my case of `WillPopScope` use in my project is like this:

```
onWillPop: () async {
  Navigator.pop(context, _content);
  return false;
},
```

I want my widget always return `_content`. How to achieve it using `PopScope`?

Another one is like this:

```
onWillPop: () async {
  if (_isCanSubmit && _selected?.path != _recordPath) return _deleteRecord();
  return true;
},
```

I need to check and call `_deleteRecord();` whenever it is needed when the widget will pop. How to achieve it using `PopScope`?

Thank you in advance.

**Comment by mkucharski17** [comment]
> Not sure about your specific case, but in general you can definitely prevent a system back from closing the app with `PopScope(canPop: false, ...)`. You should open an issue if you've got a bigger problem or you suspect a bug in PopScope.

Thank you for the fast answer :)

It is probably an issue with my project. I tried to reproduce it in a simple example based on the default app created with `flutter create` and `PopScope` worked as it should. Exiting using the native back button on Android was blocked.

**Comment by justinmc** [comment]
@mkucharski17 Great! Let me know if it comes up again or seems to not be working right.

**Comment by TatsuUkraine** [comment]
> Or is it really needed to deprecate `WillPopScope`? Are they cannot both co-exist? Why?

It seems with PopScope you can achieve the same results, but in a little bit different way. But maybe WillPopScope can be just refactored to use PopScope inside, since I can assume a lot of people will be doing the same work with canPop:false and manual pop in the callback to mimick WillPopScope functionality 

> Can anyone or Flutter team itself explain how to migrate from `WillPopScope` to `PopScope` in a video of anything which is easy to understand?

If it helps, there is a migration doc https://docs.flutter.dev/release/breaking-changes/android-predictive-back that has basic examples

**Comment by NeariX67** [comment]
No, `PopScope` does not achieve the same results as `WillPopScope`.
With `WillPopScope.onWillPop` you could ask the user if he is sure about leaving the dialog when changes are not yet saved. This was possible because the `onWillPop` Function was asynchronous and it returned a bool whether it should pop or not.

With `PopScope` you don't have the possibility to cancel the popping action based on user choice, because `canPop` is neither a function nor asynchronous. And it's already too late for `onPopInvoked`.

I hope this make it clear that this specific use case of WillPopScope will no longer be achievable with PopScope.

**Comment by TatsuUkraine** [comment]
> With `PopScope` you don't have the possibility to cancel the popping action based on user choice, because `canPop` is neither a function nor asynchronous. And it's already too late for `onPopInvoked`.

https://docs.flutter.dev/release/breaking-changes/android-predictive-back#migrating-a-back-confirmation-dialog

**Comment by justinmc** [comment]
@topex-psy Responding to your two questions about how to migrate:

### Returning a value from a system back gesture
You can do this, but it might be a bad idea. System back gestures return `null` and Flutter doesn't provide a supported way to override that. You probably want to find another way to get that value to the parent route, like using some form of state management for example.

That said, the following example, works, but use it at your own risk. If you needed to use this trick in a root route for some reason, Android predictive back wouldn't work when swiping back to leave the app. In a non-root route, it may disable [predictive back transitions](https://github.com/flutter/flutter/issues/131961) in the future for that route.

https://gist.github.com/justinmc/c436b8d82136ec77c8026aa5318da804
https://dartpad.dev/?id=c436b8d82136ec77c8026aa5318da804

### Making a call on pop
`PopScope.onPopInvoked` will be called when the route is popped, whether from a system back gesture or `Navigator.pop`. It is called just *after* the pop happens though. Does something like this work in your case?:

```dart
PopScope(
  canPop: true,
  onPopInvoked: (bool didPop) {
    if (!didPop) {
      return;
    }
    // The route has been popped.
  },
)
```

**Comment by florian-obernberger** [comment]
> With `WillPopScope.onWillPop` you could ask the user if he is sure about leaving the dialog when changes are not yet saved. This was possible because the `onWillPop` Function was asynchronous and it returned a bool whether it should pop or not.

I am using `PopScope` for exactly that purpose and besides the aforementioned issues with go_router it's working flawlessly! :)

**Comment by topex-psy** [comment]
> That said, the following example, works, but use it at your own risk. If you needed to use this trick in a root route for some reason, Android predictive back wouldn't work when swiping back to leave the app. In a non-root route, it may disable [predictive back transitions](https://github.com/flutter/flutter/issues/131961) in the future for that route.
> 
> https://gist.github.com/justinmc/c436b8d82136ec77c8026aa5318da804 https://dartpad.dev/?id=c436b8d82136ec77c8026aa5318da804

What a comprehensive answer and explanation. Yeah, I know this is not a good practice, but in my case it's too complex and would be too difficult to use state management or other means, so I took my risk since the beginning, no errors reported on that nevertheless.

Looking at the example you provided, now I have a better understanding of how to handle several `WillPopScope` migrations in my project. It's very helpful. Let's see if I can complete my migration.

**Comment by topex-psy** [comment]
> ### Making a call on pop
> `PopScope.onPopInvoked` will be called when the route is popped, whether from a system back gesture or `Navigator.pop`. It is called just _after_ the pop happens though. Does something like this work in your case?:
> 
> ```dart
> PopScope(
>   canPop: true,
>   onPopInvoked: (bool didPop) {
>     if (!didPop) {
>       return;
>     }
>     // The route has been popped.
>   },
> )
> ```

I need to verify, are these correct?

Before:
```
WillPopScope(
  onWillPop: _onWillPop,
)
```
After:
```
PopScope(
  canPop: false,
  onPopInvoked: (didPop) async {
    if (didPop) return;
    if (await _onWillPop()) Navigator.pop(context);
  },
)
```

---

Before:
```
WillPopScope(
  onWillPop: () async {
    forYouGetAllData?.call();
    return true;
  },
)
```
After:
```
PopScope(
  canPop: true,
  onPopInvoked: (didPop) {
    if (didPop) return;
    forYouGetAllData?.call();
  },
)
```

**Comment by ziqq** [comment]
`PopScope` not working in `ModalBottomSheet` for me.

```dart
PopScope(
  canPop: false,
  onPopInvoked: (didPop) async {
    if (didPop) return;
    if (await _onWillPop(context, model)) {
      if (!context.mounted) return;
      Navigator.of(context).pop();
    }
  },
  child: child,
);
```

**Comment by iainsmacleod** [comment]
Here is an example of my solution, for what it's worth - FYI, I got details from here:

https://docs.flutter.dev/release/breaking-changes/android-predictive-back

class MyScreen extends StatelessWidget {
  Future<bool?> _showBackDialog(BuildContext context) async {
    return showDialog<bool>(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          title: Text("Discard Edit?"),
          content: Text("Are you sure you want to cancel edit?"),
          actions: [
            TextButton(
              onPressed: () {
                Navigator.of(context).pop(true); // Confirm discard
              },
              child: Text(
                "Yes",
                style: Theme.of(context).textTheme.bodyLarge?.merge(
                      TextStyle(color: Colors.red),
                    ),
              ),
            ),
            TextButton(
              onPressed: () {
                Navigator.of(context).pop(false); // Cancel discard
              },
              child: Text(
                "No",
                style: Theme.of(context).textTheme.bodyLarge,
              ),
            ),
          ],
        );
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('My Screen'),
      ),
      body: PopScope(
        canPop: false,
        onPopInvoked: (bool didPop) async {
          if (didPop) {
            return;
          }
          final NavigatorState navigator = Navigator.of(context);
          final bool? shouldPop = await _showBackDialog(context);
          if (shouldPop ?? false) {
            navigator.pop();
          }
        },
        child: Center(
          child: Text('Press the back button to show confirmation dialog.'),
        ),
      ),
    );
  }
}

void main() {
  runApp(MaterialApp(
    home: MyScreen(),
  ));
}

**Comment by emirhalici** [comment]
There is confusion around the new widget and the [migration guide](https://docs.flutter.dev/release/breaking-changes/android-predictive-back). Because it isn't handling most common use cases we are confused as if this change has removed functionality. Technically this widget below should preserve the same previous api:

```dart
// ignore: deprecated_member_use
/// Replicates behaviour of deprecated [WillPopScope]. Implemented to provide backwards compatibility while
/// migrating off of it. Warning! USING THIS WIDGET WILL DISABLE PREDICTIVE BACK ON ANDROID AND BACK SWIPES ON IOS.
class CustomWillPopScope extends StatelessWidget {
  const CustomWillPopScope({
    required this.child,
    required this.onWillPop,
    super.key,
  });

  /// The widget below this widget in the tree.
  ///
  /// {@macro flutter.widgets.ProxyWidget.child}
  final Widget child;

  /// Decides whether do pop the current route. Internally disables back swipes and Android predictive back feature.
  final Future<bool> Function()? onWillPop;

  @override
  Widget build(BuildContext context) {
    if (onWillPop == null) {
      return child;
    } else {
      return PopScope(
        // Always disable predictive back feature
        canPop: false,
        child: child,
        onPopInvoked: (didPop) async {
          if (didPop) {
            // Do nothing, if system already popped the current route, re-popping might remove main route
          } else {
            // Let [GemWillPopScope] user decide whether to pop
            final bool shouldActuallyPop = await onWillPop!();
            if (shouldActuallyPop) popPage(context);
          }
        },
      );
    }
  }
}
```

The common use case was: when leaving the page pop back with the value. The change would be from this:
```dart
class ExampleWidget extends StatelessWidget {
  const ExampleWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        const result = 1;
        // We manually pop
        Navigator.of(context).pop(result);
        // Telling widget to not pop
        return false;
      },
      child: Scaffold(
        appBar: AppBar(),
        body: const Text('Hello world'),
      ),
    );
  }
}
```
to
```dart
class ExampleWidget2 extends StatelessWidget {
  const ExampleWidget2({super.key});

  @override
  Widget build(BuildContext context) {
    return PopScope(
      // Disable predictive back; but does not disable attemps
      canPop: false,
      onPopInvoked: (didPop) {
        if (didPop) {
          // Flutter already popped the page
        } else {
          // We can decide to manually pop the page
          const result = 1;
          Navigator.of(context).pop(result);
        }
      },
      child: Scaffold(
        appBar: AppBar(),
        body: const Text('Hello world'),
      ),
    );
  }
}
```

I think this should've been better documented and it took me a while to understand the real implications of the change in the api.

**Comment by ziqq** [comment]
Not working in **Flutter** `3.24`

```dart
PopScope(
  canPop: false,
    onPopInvokedWithResult: (didPop, result) async {
      if (didPop) {
        // Do somthing...
      } else {
        final navigator = Navigator.of(context);
        final bool shouldClose = await _onWillPop(context, viewModel);
        if (shouldClose) navigator.pop(result);
      }
  },
  child: const SizedBox.shrink(),
)
```

```dart
Future<bool> _onWillPop(BuildContext context) async {
  if (!mounted) return false;
  if (viewModel.cantWillPop) {
    HapticFeedback.vibrate().ignore();
    return await UiKit.showCupertinoDialog<bool>(
      context: context,
      barrierDismissible: false,
      title: context.stringOf().textAttention,
      description: context.stringOf().textWarningAboutDataLossWhenClosingTheWindow,
      actions: [
        CupertinoDialogAction(
          isDefaultAction: true,
          onPressed: () {
            Navigator.of(context, rootNavigator: true).pop(false);
          },
          child: Text(
            context.stringOf().actionCancel,
            style: Theme.of(context)
                .textTheme
                .bodyLarge
                ?.copyWith(color: Theme.of(context).colorScheme.secondary),
          ),
        ),
        CupertinoDialogAction(
          isDestructiveAction: true,
          onPressed: () {
            Navigator.of(context, rootNavigator: true).pop(true);
          },
          child: Text(context.stringOf().actionClose),
        ),
      ],
    ).then<bool>((result) => result ?? false);
  } else {
    return Future.value(true);
  }
}
```

**Comment by roscadalexandru** [comment]
any fix for this? I have the same problem with the modal. I can prevent tapping outside the modal, but if I drag it down the end it bypasses the prevention

**Comment by ziqq** [comment]
> any fix for this? I have the same problem with the modal. I can prevent tapping outside the modal, but if I drag it down the end it bypasses the prevention

This has not been fixed yet

**Comment by delfme** [comment]
Please don’t discontinue it, we rely on it heavily, and PopScope isn’t a viable alternative for our use case.

**Comment by TatsuUkraine** [comment]
Idk, you can fully replace willpopscope with popscope, especially if you follow migration guide. If the popscope doesn't work like doc says, it's not a reason to keep willpopscope instead of fix popscope. We successfully migrated as soon as willpopscope become deprecated and never had issues with popscope

**Comment by TatsuUkraine** [comment]
I need to fix myself in above comment) we "almost never had problems") but it's simply because popscope has fixable issues, which is not the reason to keep willpopscope, especially if take into account that it's incompatible with android predictable back navigation

**Comment by delfme** [comment]
@TatsuUkraine https://github.com/flutter/flutter/issues/163052

**Comment by TatsuUkraine** [comment]
@delfme I can be wrong thee, and misread the issue that you referring to. It seems like user in the original issue just not using popscope like it's designed to be used. Unlike WillPopScope which is blocks back navigation in the callback, PopScope blocks back navigation using imperative way, where you pass "false" to canPop, and then decide if you actually want to pop current route programmatically in the callback. Or passing by true to canPop, callback becomes more like a notifier rather than a stopper.

But again, I may misread the issue that you referring to. In our case I was able successfully get rid off WillPopScope in favor of PopScope. And recreating original behavior of the WillPopScope using PopScope widget

**Comment by TatsuUkraine** [comment]
Also, there is (or was) some ongoing issues with router packages such as go_router that didn't work well with Android predictive navigation in a first place (I this there was a comment somewhere in this thread listing opened issues that popscope has currently)

**Comment by TatsuUkraine** [comment]
The point which I'm trying to make is that, based on conversation it looks to me that the main issue is that PopScope doesn't work like it should be, and it doesn't mean that original WillPopScope should stay. It just means that PopScope should be adjusted to the point where it will actually be working like it should be
